### PR TITLE
SNOW-1458131: Add support for Index.is_unique and Index.has_duplicates

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1458131

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Add support for Index.is_unique and Index.has_duplicates

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_internal/frame.py [hunk 1]
```diff
@@ -609,16 +616,30 @@ def has_unique_index(self, axis: Optional[int] = 0) -> bool:
...
             return self.data_columns_index.is_unique
         else:
-            # Note: We can't use 'count_distinct' because it ignores null values.
-            total_rows = self.num_rows
-            distinct_rows = count_rows(
-                get_distinct_rows(
-                    self.ordered_dataframe.select(
-                        self.index_column_snowflake_quoted_identifiers
+            # Implement using single query for single index.
+            if self.num_index_columns == 1:
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/_internal/frame.py:620]: why can't we do the same for multiple index columns?

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/frame.py:620]: For multiple index columns expression will be very complex. Happy to update the code if you have any suggestion for the expression.

File: src/snowflake/snowpark/modin/plugin/_internal/frame.py [hunk 2]
```diff
@@ -609,16 +616,30 @@ def has_unique_index(self, axis: Optional[int] = 0) -> bool:
...
+                # COUNT(DISTINCT) ignores NULL values, so if there is a NULL value in the column,
+                # we include it via IFF(MAX(<col> IS NULL)), 1, 0) which will return 1 if there is
+                # at least one NULL contained within a column, and 0 if there are no NULL values.
+                rows = self.ordered_dataframe._dataframe_ref.snowpark_dataframe.select(
+                    (
+                        count_distinct(index_col) + iff(max_(index_col.is_null()), 1, 0)
+                    ).as_("unique_count"),
+                    count("*").as_("total_count"),
+                ).collect()
+                return len(rows) == 0 or rows[0][0] == rows[0][1]
```
[the remark this answers is not in the document]
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/frame.py]: removed it.

File: tests/integ/modin/index/test_index_methods.py [hunk 3]
```diff
@@ -347,3 +352,29 @@ def test_df_index_columns_dtype(native_df):
     # do not check columns type for empty df
     if not native_df.empty:
         assert snow_df.columns.dtype == native_df.columns.dtype
+
+
+UNIQUE_TEST_DATA = NATIVE_INDEX_UNIQUE_TEST_DATA + [
```
[the remark this answers is not in the document]
Comment (sfc-gh-nkumar) [reply tests/integ/modin/index/test_index_methods.py]: Was seeing weird behavior in other methods with multi-index. I learned from Varnika that this class not supposed to handle multi-index. Removed multi-index tests.

File: src/snowflake/snowpark/modin/plugin/_internal/frame.py (hunk 1 above)
Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/_internal/frame.py:620]: I think this works:

```python
            return self.ordered_dataframe._dataframe_ref.snowpark_dataframe.select(
                count_distinct(array_construct(*self.index_column_snowflake_quoted_identifiers)) == count("*"),
            ).collect()[0][0]
```

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/frame.py:620]: interesting idea, let me try this.

File: src/snowflake/snowpark/modin/plugin/_internal/frame.py [hunk 4]
```diff
@@ -609,16 +617,25 @@ def has_unique_index(self, axis: Optional[int] = 0) -> bool:
...
-            distinct_rows = count_rows(
-                get_distinct_rows(
-                    self.ordered_dataframe.select(
-                        self.index_column_snowflake_quoted_identifiers
+            if self.num_index_columns == 1:
+                index_col = col(self.index_column_snowflake_quoted_identifiers[0])
+                # COUNT(DISTINCT) ignores NULL values, so if there is a NULL value in the column,
+                # we include it via IFF(MAX(<col> IS NULL), 1, 0) which will return 1 if there is
+                # at least one NULL contained within a column, and 0 if there are no NULL values.
+                return self.ordered_dataframe._dataframe_ref.snowpark_dataframe.select(
```
Comment (sfc-gh-vbudati) [inline src/snowflake/snowpark/modin/plugin/_internal/frame.py:625]: I think it might be better to move this logic to the query compiler layer

**Review by sfc-gh-vbudati — APPROVED** [review]
Looks good, thanks Naresh!

**Review by sfc-gh-mvashishtha — APPROVED** [review]
thanks!

**Pull request merged**